contracts/interchain-token-service/src/contract/execute/mod.rs

```diff
@@ -157,10 +173,76 @@ fn apply_to_hub(
             apply_to_token_deployment(storage, &source_chain, &destination_chain, deploy_token)
                 .map(Message::DeployInterchainToken)?
         }
+        Message::RegisterToken(register_token) => {
+            apply_to_register_token(storage, source_chain, register_token)
+                .map(Message::RegisterToken)?
+        }
+        Message::LinkToken(link_token) => {
+            apply_to_link_token(storage, source_chain, destination_chain, link_token)
+                .map(Message::LinkToken)?
+        }
     }
     .then(Result::Ok)
 }
 
+fn apply_to_register_token(
+    storage: &mut dyn Storage,
+    source_chain: ChainNameRaw,
+    register_token: RegisterToken,
+) -> Result<RegisterToken, Error> {
+    interceptors::register_custom_token(storage, source_chain, register_token.clone())?;
+    Ok(register_token)
+}
+
+fn apply_to_link_token(
+    storage: &mut dyn Storage,
+    source_chain: ChainNameRaw,
+    destination_chain: ChainNameRaw,
+    link_token: LinkToken,
+) -> Result<LinkToken, Error> {
+    let source_token = state::may_load_custom_token(
+        storage,
+        source_chain.clone(),
+        link_token.source_token_address.clone(),
+    )
+    .change_context(Error::State)?
+    .ok_or(Error::TokenNotRegistered(
+        link_token.source_token_address.clone(),
+    ))?;
+
+    deploy_token_to_source_chain(
+        storage,
+        &source_chain,
+        link_token.token_id,
+        source_token.decimals,
+    )?;
+
+    let destination_decimals = if link_token.autoscaling {
+        state::may_load_custom_token(
+            storage,
+            destination_chain.clone(),
+            link_token.destination_token_address.clone(),
+        )
+        .change_context(Error::State)?
+        .ok_or(Error::TokenNotRegistered(
+            link_token.destination_token_address.clone(),
+        ))?
+        .decimals
+    } else {
+        source_token.decimals
```

Comment:
Correct me if I am wrong, say chain A registers token T with X decimals and chain B registers token T with Y decimals, and then we deploy link token T from chain A to chain B with autoscaling disabled. Now token T has both X decimals for chain A & B. If now chain C also registers token T, it can never be linked from chain B but has to be from chain A because `deploy_token_to_source_chain` is gonna see a decimal mismatch of Token T on chain B. Is this an issue?

Reply:
`deploy_token_to_source_chain` will prevent linking from a different chain other than A since a global token config is stored for that tokenId after the first deployment and the method ensures the source chain for LinkToken doesn't change. And ITS edge won't allow linking from a different origin chain since tokenId derivation will be different